### PR TITLE
REF: document casting behavior in groupby

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -967,11 +967,22 @@ def _cython_operation(
         )
 
     @final
-    def agg_series(self, obj: Series, func: F) -> ArrayLike:
+    def agg_series(self, obj: Series, func: F, preserve: bool = False) -> ArrayLike:
```
Comment (jreback) [inline pandas/core/groupby/ops.py:970]: preserve_dtype? I don't know if we use this keyword elsewhere

Comment (jreback) [reply pandas/core/groupby/ops.py:970]: or cast?

**Review by jreback — CHANGES_REQUESTED** [review]
also pls rebase

**Comment by jreback** [comment]
and perf question

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:970]: changing to preserve_dtype

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1236,12 +1233,11 @@ def _wrap_applied_output_series(
             columns = key_index
             stacked_values = stacked_values.T
 
+        if stacked_values.dtype == object:
+            # We'll have the DataFrame constructor do inference
+            stacked_values = stacked_values.tolist()
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py:1238]: test_apply_numeric_coercion_when_datetime goes through here.  running the usage in timeit looks like a wash

**Comment by jbrockmendel** [comment]
updated + green

**Review by jreback — APPROVED** [review]

File: pandas/core/groupby/generic.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py:1253]: i think we may be able to do better/cleaner following #40489

**Pull request merged**